File: c/sender.c
/*
SENDER
Please read "Dokumentation_und_Auswertung.pdf" for more precise information
uses zlib.h
compile with:
gcc -o sender sender.c -lz
*/
#include <sys/types.h>
#include <sys/socket.h>
#include <stdio.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <stdlib.h>
#include <strings.h>
#include <string.h>
#include <zlib.h>
#include <unistd.h>
#include "sender.h"

int main(int argc, char *argv[]){


    //handle wrong number of cmd args
    if(argc != 7){
        handle_error("Usage: ./sender <receiver name> <port> <number of packets to send> <packet payload size> <ack-timeout> <window-size>");
    }

    //declare and define required local vars
    int LAR;
    int sock, amt_send, amt_sent, sent, total_pack_size, payload, sockaddr_len, i;
    struct sockaddr_in receiver, sender;
    struct hostent *host;
    struct timeval before, after;
    struct timeval ack_timeout, sleep;
    uint crc;
    crc = 0;
    payload = atoi(argv[4]);
    total_pack_size = payload + 1;
    int ack_buffer[1];
    int buffer[total_pack_size];
    for(i = 0; i < total_pack_size; i++){
        buffer[i] = htonl(9);
    }
    LAR = -1;
    amt_send = atoi(argv[3]);
    int window_size;
    window_size = atoi(argv[6]);

    //create udp ipv4 socket and handle creation error
    sock = socket(AF_INET, SOCK_DGRAM, 0);
    if(sock < 0){
        handle_error("socket could not be created");
    }

    /*set after first receive such that the receiver waits for the first packet
    indefinitely*/
    ack_timeout.tv_sec = 0;
    ack_timeout.tv_usec = atoi(argv[5]) * 1000;
    if(setsockopt(sock,SOL_SOCKET,SO_RCVTIMEO,&ack_timeout,sizeof(ack_timeout)) < 0){
        handle_error("could not set socket receive-timeout");
    }


    //set sender info: protocol and host and handle host creation error
    receiver.sin_family = AF_INET;
    host = gethostbyname(argv[1]);
    if(host == 0){
        handle_error("host not recognized");
    }

    //copy host address into sockaddr_in struct -> add to addr info and set port
    bcopy((char *)host->h_addr, (char *)&receiver.sin_addr, host->h_length);
    receiver.sin_port = htons(atoi(argv[2]));
    gettimeofday(&before, NULL);
    //start sending
    for(amt_sent = 0; amt_sent < amt_send; amt_sent++){
        //set sequence number
        buffer[0] = htonl(amt_sent);
        //check cases
        if(amt_sent < amt_send - 1){
            crc = crc32(crc, (const void*)buffer, total_pack_size);
        }else{
            crc = crc32(crc, (const void*)buffer, total_pack_size - 1);
            buffer[total_pack_size - 1] = htonl(crc);
        }
        //send message in buffer as datagram and handle error
        sent = sendto(sock, buffer, sizeof(buffer), 0, (struct sockaddr *)&receiver, sizeof(struct sockaddr_in));
        if(sent == 0){
            handle_error("package not sent successfully");
        }
      /*handle acks, go-back-n if either ack not expected ack or no ack received for
      first packet in sliding window */
      if(amt_sent == LAR + window_size){
          if(recvfrom(sock,ack_buffer,sizeof(ack_buffer),0,(struct sockaddr*)&sender,&sockaddr_len) >= 0){
              if(ntohl(ack_buffer[0]) == LAR + 1){
                  LAR = ntohl(ack_buffer[0]);
              }else{
                amt_sent = LAR;
              }
          }else{
            amt_sent = LAR;
          }
      }
    }
    gettimeofday(&after, NULL);
    evaluate(before,after,total_pack_size,amt_send,crc);
    return 0;
}

/*
evaluates the send-operation. calculates transfer-speed and prints it.
*/
void evaluate(struct timeval before, struct timeval after, int msg_size, int amt, uint crc){
    time_t duration = (after.tv_usec - before.tv_usec) +(after.tv_sec - before.tv_sec)*1000000;
    long total_size = msg_size * amt * 4;
    printf("crc32-checksum: %u\n", crc);
    printf("%d packets sent\n", amt);
    if(duration != 0.0){
        double speed = 8 * total_size/duration;
        printf("%.3f mbit/s\n", speed);
    }else{
        handle_error("measured time too short. Try sending more packets");
    }
}

void handle_error(char *message){
    puts(message);
    exit(1);

}

File: java/src/Sender.java
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketTimeoutException;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.zip.CRC32;

/**
 * Sender represents a UDP packet sender. Parameters can be specified via cmd.
 * Packets can be sent to a single recipient. A checksum over all Packets is and
 * sent as part of the last packet.
 * Uses a sliding-window to ensure no packets are lost.
 * @author Thomas Samy Dafir, 1331483
 * @author Dominik Baumgartner, 0920177
 * @author Vivien Wallner, 1320293
 */
public class Sender {
	
	private static int LAR;
	private static int windowSize;
	
	/**
	 * main handles wrong numbers of cmd-arguments.
	 * Receiver-name, receiver-port, sending-steps and the amount of packets to send in each step have to be 
	 * specified, otherwise a message is issued informing the user on how to use the application. IOExceptions
	 * thrown in the startSending-method are handled here.
	 * @param args cmd-arguments:
	 * 	args[0] receiver name or IP-address
 	 * 	args[1] receiver port
 	 * 	args[2] number of packets to send
 	 *  args[3] size of each packet's User-Message 	
	 */
	public static void main(String[] args){
		LAR = -1;
		if(args.length != 5){
			System.out.println("Usage: java Sender <receiver name> <port> <number of packets to send> <payload size> <window-size>");
		}else{
			try{
				startSending(args);
			}catch(IOException e){
				System.err.println("issue creating socket");
			}catch(InterruptedException e){}
		}
	}
	/**
	 * startSending handles sending of UDP packets. Creates a socket on free port (sender-port-number
	 * is not important in our application since the receiver does not send an answer to the sender).
	 * Gets the receiver's IP-address from args[0] and sends packets to the 
	 * receiver (sending process is split into args[2] steps).
	 * in each step the number of packets specified in args[3,4,...] is sent.
	 * @param args cmd-arguments passed from main
	 * @throws IOException exception is forwarded to be handled by calling method
	 * @throws InterruptedException 
	 */
	private static void startSending(String[] args) throws IOException, InterruptedException{
        long beforeTime = 0;
        long afterTime = 0;
        windowSize = Integer.parseInt(args[4]);
		//create socket with standard constructor. Sender-port does not have to be known.
		DatagramSocket socket = new DatagramSocket();
		//set timeout for ACK-reception
		socket.setSoTimeout(100);

		//transform the cmd-argument receiver-name into an InetAddress object.
		CRC32 checksum = new CRC32();
		InetAddress address = InetAddress.getByName(args[0]);
		int receiverSocket = Integer.parseInt(args[1]);
		int size = Integer.parseInt(args[3]) * 4;
		ByteBuffer buffer = ByteBuffer.allocate(size + 4);
		ByteBuffer ackBuffer = ByteBuffer.allocate(4);
		byte[] message = getPayload(size);
		//DatagramPacket checkPacket = new DatagramPacket(buffer, buffer.length);
		int sendAmt = Integer.parseInt(args[2]);
		DatagramPacket sendPacket;
		DatagramPacket ackPacket = new DatagramPacket(ackBuffer.array(), ackBuffer.array().length);
		
		//the for-loop handles sending of packets. creates a string-message including the packet-number.
		//Create byte-array from string, receiver-info to generate packet and
		//send the packet using the datagram socket.
		beforeTime = System.currentTimeMillis();
		int ackNumber;
		for(int i = 0; i < sendAmt; i++){
			buffer.putInt(i);
			/*If current packet is not the last packet: put the whole message into the bytebuffer and use it
			 * to update the checksum. Otherwise: put message up to length - 4 into the buffer, to leave space
			 * for the checksum + add checksum.*/
			if(i < sendAmt - 1){
				buffer.put(message);
				checksum.update(buffer.array());
			}else{
				buffer.put(Arrays.copyOfRange(message, 0, message.length-4));
				checksum.update(Arrays.copyOfRange(buffer.array(), 0, buffer.array().length-4));
				buffer.putInt((int) checksum.getValue());
			}
			sendPacket = new DatagramPacket(buffer.array(), buffer.array().length, address, receiverSocket);
			socket.send(sendPacket);
			if(i == LAR + windowSize){
				try{
					socket.receive(ackPacket);
					ackBuffer = ByteBuffer.wrap(ackPacket.getData());
					ackNumber = ackBuffer.getInt();
					if(ackNumber != LAR + 1){
						System.out.println(i + " wrong packet " + ackNumber + " " + LAR);
						i = LAR;
						Thread.sleep(10);
					}else{
						LAR = ackNumber;
					}
				}catch(SocketTimeoutException e){
					System.out.println("timeout");
					i = LAR;
					Thread.sleep(10);
				}
				
			}
			buffer.clear();
		}
		afterTime = System.currentTimeMillis();
		socket.close();
		//+4 for one integer --> sequence number
		evaluate(afterTime, beforeTime, size + 4, sendAmt);
		System.out.println("checksum: " + checksum.getValue());
	}
	
	/**
	 * constructs the User-Data for to be sent
	 * @param length length of result string
	 * @return string of given length
	 */
	private static byte[] getPayload(int length){
		ByteBuffer bb = ByteBuffer.allocate(length);
		for(int i = 0; i < length/4; i++){
			bb.putInt(9);
		}
		return bb.array();
		
	}
	
	/**
	 * evaluates the sending process. Computes Transfer-speed based on sent data and transfer-time
	 * @param after timestamp after sending
	 * @param before timestamp before sending
	 * @param size size of a package's Message
	 * @param amnt Amount of packets sent
	 */
	private static void evaluate(long after, long before, int size, int amnt){
		long duration = after - before;
		int totalSize = size * amnt;
		double speed = (8 * (double)(totalSize)/(double)(duration)) / 1000;
		System.out.println(String.format("%.3f mbit/s", speed));		
	}
}